### PR TITLE
Laboratory work #1, Elena Chevtaeva - 23FPL2

### PR DESCRIPTION
**Pull request by Helenhelen2005**

### CODE REVIEW

**Comment by Helenhelen2005** [comment]
add an empty line

**Comment by Vasilisa-Blyudova** [comment]
Nothing to review today. 
I advise you to start working as soon as possible! 
Good luck!

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -20,6 +20,15 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if isinstance(text, str) and len(text) > 0:
+        clean_text = ""
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:24]: create a list instead of a string

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -34,6 +43,25 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if tokens is None or not isinstance(tokens, list):
+        return None
+
+    freq_dict = {}
+    new_tokens = []
+
+    for i in tokens:
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:52]: rename variable

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -34,6 +43,25 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        if isinstance(i, str) and len(i) == 1:
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:53]: return None

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -49,6 +77,15 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if (language is not None and isinstance(language, str)
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:80]: return None

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -64,6 +101,12 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if isinstance(predicted, list) and isinstance(actual, list):
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:104]: return None here

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -84,6 +127,34 @@ def compare_profiles(
     In case of corrupt input arguments or lack of keys 'name' and
     'freq' in arguments, None is returned
     """
+    if isinstance(unknown_profile, dict) and isinstance(profile_to_compare, dict):
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:130]: return None here

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -84,6 +127,34 @@ def compare_profiles(
...
+        name1 = unknown_profile.get('name')
+        freq1 = unknown_profile.get('freq')
+        name2 = profile_to_compare.get('name')
+        freq2 = profile_to_compare.get('freq')
+
+        if (isinstance(name1, str) and isinstance(freq1, dict)
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:136]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -84,6 +127,34 @@ def compare_profiles(
...
+                actual_value = freq1.get(token, 0.0)
+                actual_values.append(actual_value)
+
+                predicted_value = freq2.get(token, 0.0)
+                predicted_values.append(predicted_value)
+
+            mse = calculate_mse(actual_values, predicted_values)
+
+            if mse is not None:
+                final_mse = round(mse, 3)
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:152]: no need to round

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -105,6 +176,26 @@ def detect_language(
 
     In case of corrupt input arguments, None is returned
     """
+    if unknown_profile is not None and profile_1 is not None and profile_2 is not None:
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:179]: the same

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Good job! Raise your score to 10!

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:81]: I have "return None" at the end of the cycle. I tried to place it in a different way, but got mistakes with returns (there was something like  "you cannot return value and None").

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:105]: I have "return None" at the end of the cycle. I tried to place it in a different way, but got mistakes with returns (there was something like  "you cannot return value and None").

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:131]: I have "return None" at the end of the cycle. I tried to place it in a different way, but got mistakes with returns (there was something like  "you cannot return value and None").

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:137]: I have "return None" at the end of the cycle. I tried to place it in a different way, but got mistakes with returns (there was something like  "you cannot return value and None").

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:153]: I do not really understand why, in the task we have "Ее значение будет равно 0.167 (с округлением до третьего знака)." Is it only rounded in the task to make the number easier to read for us?

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:180]: I have "return None" at the end of the cycle. I tried to place it in a different way, but got mistakes with returns (there was something like  "you cannot return value and None").

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -3,6 +3,9 @@
 
 Language detection
 """
+import json
+
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:7]: rm whitespaces

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -20,6 +23,14 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if isinstance(text, str) and len(text) > 0:
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:26]: return None here

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (Vasilisa-Blyudova) [reply lab_1_classify_by_unigrams/main.py:84]: None returns if, for example, language isn't string

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (Vasilisa-Blyudova) [reply lab_1_classify_by_unigrams/main.py:108]: .

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (Vasilisa-Blyudova) [reply lab_1_classify_by_unigrams/main.py:134]: .

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (Vasilisa-Blyudova) [reply lab_1_classify_by_unigrams/main.py:140]: .

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (Vasilisa-Blyudova) [reply lab_1_classify_by_unigrams/main.py:182]: .

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -149,11 +281,26 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
+                return None
+            collected_profiles.append(new_profile)
+
+    return collected_profiles
 
 
 def detect_language_advanced(
-    unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
-) -> list | None:
+            unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:302]: rm whitespaces

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -167,6 +314,27 @@ def detect_language_advanced(
 
     In case of corrupt input arguments, None is returned
     """
+    if unknown_profile is not None and known_profiles is not None:
+        if isinstance(unknown_profile, dict) and isinstance(known_profiles, list):
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:318]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -167,6 +314,27 @@ def detect_language_advanced(
...
+
+            for profile in known_profiles:
+                if isinstance(profile, dict) and all(key in profile for key in ['name', 'freq']):
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:321]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -178,3 +346,8 @@ def print_report(detections: list[tuple[str, float]]) -> None:
 
     In case of corrupt input arguments, None is returned
     """
+    if detections is not None and isinstance(detections, list):
+        for element in detections:
+            if isinstance(element, tuple):
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:351]: the same

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great! Fix the comments and get ready for the final review.

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 16]
```diff
@@ -20,6 +23,14 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if isinstance(text, str) and len(text) > 0:
+        tokens = []
+        for symbol in text.lower().strip():
+            if symbol.isalpha():
```
[the remark this answers is not in the document]
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 17]
```diff
@@ -49,6 +81,15 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if (language is not None and isinstance(language, str)
+            and text is not None and isinstance(text, str)):
+        tokenized_text = tokenize(text)
```
[the remark this answers is not in the document]
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -119,6 +213,13 @@ def load_profile(path_to_file: str) -> dict | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if isinstance(path_to_file, str):
```
[the remark this answers is not in the document]
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 19]
```diff
@@ -135,6 +236,37 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
     In case of corrupt input arguments or lack of keys 'name', 'n_words' and
     'freq' in arguments, None is returned
     """
+    if profile is not None:
+        if (isinstance(profile, dict)
```
[the remark this answers is not in the document]
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 20]
```diff
@@ -149,11 +281,26 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
     """
+    if not isinstance(paths_to_profiles, list):
+        return None
+
+    collected_profiles = []
+
+    for path in paths_to_profiles:
+        if load_profile is None and preprocess_profile is None:
+            return None
+        if isinstance(path, dict):
```
[the remark this answers is not in the document]
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:7]: done

File: lab_1_classify_by_unigrams/main.py [hunk 21]
```diff
@@ -34,6 +44,27 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
     In case of corrupt input arguments, None is returned
     """
+    if tokens is None or not isinstance(tokens, list):
+        return None
+
+    freq_dict = {}
+    new_tokens = []
+
+    for token in tokens:
+        if isinstance(token, str) and len(token) == 1:
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:54]: return None here if the conditions are opposite

File: lab_1_classify_by_unigrams/main.py [hunk 22]
```diff
@@ -64,6 +108,13 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(predicted, list) or not isinstance(actual, list):
+        return None
+    if len(actual) != len(predicted):
+        return None
+    n = len(actual)
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:115]: rename

File: lab_1_classify_by_unigrams/main.py [hunk 23]
```diff
@@ -105,6 +189,29 @@ def detect_language(
...
+        return None
+    if (not isinstance(unknown_profile, dict) or not isinstance(profile_1, dict)
+            or not isinstance(profile_2, dict)):
+        return None
+
+    name1 = profile_1.get('name')
+    name2 = profile_2.get('name')
+    mse_1 = compare_profiles(unknown_profile, profile_1)
+    mse_2 = compare_profiles(unknown_profile, profile_2)
+    result1 = str(name1)
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:202]: Does this really need to be explicitly converted to a string?

File: lab_1_classify_by_unigrams/main.py [hunk 24]
```diff
@@ -105,6 +189,29 @@ def detect_language(
...
+    result2 = str(name2)
+
+    if mse_1 is None or mse_2 is None:
+        return None
+
+    if mse_1 < mse_2:
+        return result1
+    if mse_1 == mse_2:
+        profile_sorted_list = sorted([result1, result2])
+        return str(profile_sorted_list[0])
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:212]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 25]
```diff
@@ -135,6 +249,35 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+    name = profile['name']
+    freq = profile['freq']
+    n_words = profile['n_words']
+
+    total_number = n_words[0]
+    if not isinstance(total_number, int) or total_number < 1:
+        return None
+
+    freq_dict = {}
+    for k, v in freq.items():
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:269]: rename

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Excellent! Fixed the comments and get ready to submit the work!

File: lab_1_classify_by_unigrams/main.py (hunk 25 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 22 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 21 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 24 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py:212]: I don't really get how I can remove str here if the variable profile_sorted_list is a list, and I need to get str as the result of the function.

File: lab_1_classify_by_unigrams/main.py (hunk 23 above)
Comment (Helenhelen2005) [reply lab_1_classify_by_unigrams/main.py]: Actually, Mypy sees a mistake here in the type of variables if I remove str. I tried to fix this by doing explicit type check before result1 and result2. Not sure if it is what you expected me to do.

**Pull request closed without merging**